rocsp/config/rocsp_config.go

```diff
@@ -138,3 +146,80 @@ func MakeReadClient(c *RedisConfig, clk clock.Clock) (*rocsp.Client, error) {
 	})
 	return rocsp.NewClient(rdb, timeout, clk), nil
 }
+
+// A ShortIDIssuer combines an issuance.Certificate with some fields necessary
+// to process OCSP responses: the subject name and the shortID.
+type ShortIDIssuer struct {
+	*issuance.Certificate
+	subject pkix.RDNSequence
+	shortID byte
+}
+
+// LoadIssuers take a map where the keys are filenames and the values are the
+// corresponding short issuer ID. It loads issuer certificates from the given
+// files and produces a []ShortIDIssuer.
+func LoadIssuers(input map[string]int) ([]ShortIDIssuer, error) {
+	var issuers []ShortIDIssuer
+	for issuerFile, shortID := range input {
+		if shortID > 255 || shortID < 0 {
+			return nil, fmt.Errorf("invalid shortID %d (must be byte)", shortID)
+		}
+		cert, err := issuance.LoadCertificate(issuerFile)
+		if err != nil {
+			return nil, fmt.Errorf("reading issuer: %w", err)
+		}
+		var subject pkix.RDNSequence
+		_, err = asn1.Unmarshal(cert.Certificate.RawSubject, &subject)
+		if err != nil {
+			return nil, fmt.Errorf("parsing issuer.RawSubject: %w", err)
+		}
+		var shortID byte = byte(shortID)
+		for _, issuer := range issuers {
+			if issuer.shortID == shortID {
+				return nil, fmt.Errorf("duplicate shortID in config file: %d (for %q and %q)", shortID, issuer.subject, subject)
+			}
+			if !issuer.IsCA {
+				return nil, fmt.Errorf("certificate for %q is not a CA certificate", subject)
+			}
+		}
+		issuers = append(issuers, ShortIDIssuer{
+			Certificate: cert,
+			subject:     subject,
+			shortID:     shortID,
+		})
+	}
+	return issuers, nil
+}
+
+// ShortID returns the short id of an issuer. The short ID is a single byte that
+// is unique for that issuer.
+func (si *ShortIDIssuer) ShortID() byte {
+	return si.shortID
+}
+
+// FindIssuerByID returns the issuer that matches the given IssuerID or IssuerNameID.
+func FindIssuerByID(longID int64, issuers []ShortIDIssuer) (*ShortIDIssuer, error) {
+	for _, iss := range issuers {
+		if iss.NameID() == issuance.IssuerNameID(longID) || iss.ID() == issuance.IssuerID(longID) {
+			return &iss, nil
+		}
+	}
+	return nil, fmt.Errorf("no issuer found for an ID in certificateStatus: %d", longID)
+}
+
+// FindIssuerByID returns the issuer with a Subject matching the *ocsp.Response.
+func FindIssuerByName(resp *ocsp.Response, issuers []ShortIDIssuer) (*ShortIDIssuer, error) {
```

Comment:
Why is lookup by name necessary? The docs for `ocsp.CreateResponse` say that both the `RawResponderName` and the `IssuerKeyHash` fields are populated, so we can do this lookup based on matching `resp.IssuerKeyHash` with `issuer.KeyHash`, no need to store the subject at all, I don't think.

Also: the doc comment here has the wrong function name copy-pasted in.

Reply:
Good point, I'll fix this.

Reply:
I tried this and wound up reverting it. https://pkg.go.dev/golang.org/x/crypto/ocsp#Response

```
	// RawResponderName optionally contains the DER-encoded subject of the
	// responder certificate. Exactly one of RawResponderName and
	// ResponderKeyHash is set.
	RawResponderName []byte
	// ResponderKeyHash optionally contains the SHA-1 hash of the
	// responder's public key. Exactly one of RawResponderName and
	// ResponderKeyHash is set.
	ResponderKeyHash []byte
```

That's because ResponderID is a CHOICE: https://datatracker.ietf.org/doc/html/rfc6960#page-15. And in our setup, it is set to the name, not the hash.